packages/grafana-ui/src/components/BarChart/BarChart.tsx

```diff
@@ -0,0 +1,447 @@
+// Library
+import React, { useCallback, useMemo, useRef } from 'react';
+import {
+  compareDataFrameStructures,
+  DataFrame,
+  DefaultTimeZone,
+  formattedValueToString,
+  getFieldDisplayName,
+  getFieldSeriesColor,
+  TimeRange,
+} from '@grafana/data';
+import uPlot, { Axis, Scale, Series } from 'uplot';
+import { VizLayout } from '../VizLayout/VizLayout';
+import { distribute, SPACE_BETWEEN } from './distribute';
+import { Quadtree, Rect } from './quadtree';
+
+// Types
+import { VizOrientation } from '@grafana/data';
+import { Themeable } from '../../types';
+import { BarChartOptions, BarValueVisibility } from './types';
+import { useRevision } from '../uPlot/hooks';
+import { UPlotChart } from '../uPlot/Plot';
+import { UPlotConfigBuilder } from '../uPlot/config/UPlotConfigBuilder';
+import { AxisPlacement, ScaleDistribution } from '../uPlot/config';
+import { useTheme } from '../../themes';
+import { GraphNGLegendEvent, GraphNGLegendEventMode } from '../GraphNG/types';
+import { LegendDisplayMode, VizLegendItem } from '../VizLegend/types';
+import { VizLegend } from '../VizLegend/VizLegend';
+
+export interface Props extends Themeable, BarChartOptions {
+  height: number;
+  width: number;
+  data: DataFrame;
+  onLegendClick?: (event: GraphNGLegendEvent) => void;
+  onSeriesColorChange?: (label: string, color: string) => void;
+}
+
+/**
+ * @alpha
+ */
+export const BarChart: React.FunctionComponent<Props> = ({
```

Comment:
a big complex component / function, is there a way to break this up a bit, move some of the nested functions out (drawBars, drawPoints), pointWithin, walkTo etc. 

A unit test or two and 1-2 storybook stories would also help

Reply:
pulled all the core stuff out of the react component.

fixed negative value handling.

added hover integration points for adding tooltip.